### PR TITLE
[Example] RLHF end to end example

### PR DESCRIPTION
**Pull request by apbard**
merge after #1309, #1319, #1316, #1315 + rebase

Adds a complete end 2 end RLHF pipeline

### CODE REVIEW

File: examples/rlhf/train_rlhf.py [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+
+    def update(self, current: float, n_steps: int):
+        """Returns adaptively updated KL coefficient, βₜ₊₁.
+        Arguments:
+            current: The current KL value between the newest policy and the initial policy.
+        """
+        proportional_error = np.clip(current / self.target - 1, -0.2, 0.2)  # ϵₜ
+        mult = 1 + proportional_error * n_steps / self.horizon
+        self.value *= mult  # βₜ₊₁
+        return self.value
```
Comment (vmoens) [inline examples/rlhf/train_rlhf.py:64]: Should be moved to core

File: examples/rlhf/train_rlhf.py [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+    def __init__(self, init_kl_coef: float, target: float, horizon: int):
+        self.value = init_kl_coef
+        self.target = target
+        self.horizon = horizon
+
+    def update(self, current: float, n_steps: int):
+        """Returns adaptively updated KL coefficient, βₜ₊₁.
+        Arguments:
+            current: The current KL value between the newest policy and the initial policy.
+        """
```
Comment (vmoens) [inline examples/rlhf/train_rlhf.py:60]: wrong formatting

File: examples/rlhf/train_rlhf.py [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+
+    def __init__(self, init_kl_coef: float, target: float, horizon: int):
+        self.value = init_kl_coef
+        self.target = target
+        self.horizon = horizon
+
+    def update(self, current: float, n_steps: int):
+        """Returns adaptively updated KL coefficient, βₜ₊₁.
+        Arguments:
+            current: The current KL value between the newest policy and the initial policy.
```
Comment (vmoens) [inline examples/rlhf/train_rlhf.py:59]: missing args

File: examples/rlhf/train_rlhf.py [hunk 4]
```diff
@@ -0,0 +1,339 @@
...
+        """
+        proportional_error = np.clip(current / self.target - 1, -0.2, 0.2)  # ϵₜ
```
Comment (vmoens) [inline examples/rlhf/train_rlhf.py:61]: why np? We should only work with tensors no?

File: examples/rlhf/train_rlhf.py [hunk 5]
```diff
@@ -0,0 +1,339 @@
...
+):
+    """Create a function to estimate the reward via sampling.
+
+    This function creates a new function which, given a model and a dataloader, will
+    perform multiple rollouts using the model and data sampled from the dataloader then
+    average the accumulated rewards.
+
+    For debugging purposes, we also generate responses to a fixed prompt so that the
+    quality of the model can be visually assessed during training.
+    """
```
Comment (vmoens) [inline examples/rlhf/train_rlhf.py:78]: Missing args and example

File: examples/rlhf/train_rlhf.py [hunk 6]
```diff
@@ -0,0 +1,339 @@
...
+from torchrl.data.rlhf.utils import RolloutFromModel
+
+from torchrl.objectives import ClipPPOLoss
+from torchrl.objectives.value import GAE
+from tqdm import tqdm
+from transformers import GenerationConfig, GPT2Tokenizer
+from utils import get_file_logger, resolve_name_or_path, setup
+
+
+def flatten_td(td):
```
Comment (vmoens) [inline examples/rlhf/train_rlhf.py:33]: should move to utils.py

File: examples/rlhf/train_rlhf.py [hunk 7]
```diff
@@ -0,0 +1,339 @@
...
+        Arguments:
+            current: The current KL value between the newest policy and the initial policy.
+        """
+        proportional_error = np.clip(current / self.target - 1, -0.2, 0.2)  # ϵₜ
+        mult = 1 + proportional_error * n_steps / self.horizon
+        self.value *= mult  # βₜ₊₁
+        return self.value
+
+
+def create_reward_estimator(
```
Comment (vmoens) [inline examples/rlhf/train_rlhf.py:67]: Seems like something that should be a class, not a function that returns a function

File: examples/rlhf/train_rlhf.py [hunk 8]
```diff
@@ -0,0 +1,339 @@
...
+
+    @torch.no_grad()
+    def estimate_reward(model, dataloader):
+        rollout_from_model = RolloutFromModel(model, ref_model, reward_model)
+        rewards = torch.zeros(eval_iters)
+        for k in range(eval_iters):
+            batch = next(dataloader)
+            # NOTE: disable kl for evaluation
+            td = rollout_from_model.rollout_from_data(batch, kl_coef=0.0)
+            rewards[k] = td.get(("next", "reward")).sum(dim=1).mean().item()
```
Comment (vmoens) [inline examples/rlhf/train_rlhf.py:104]: why item?

File: examples/rlhf/train_rlhf.py (hunk 7 above)
Comment (vmoens) [reply examples/rlhf/train_rlhf.py:67]: Is this purely intended for logging and testing? "estimate reward" could also mean estimate reward for training.

File: examples/rlhf/train_rlhf.py [hunk 9]
```diff
@@ -0,0 +1,339 @@
...
+        sampler=SamplerWithoutReplacement(),
+        prefetch=10,
+    )
+
+    rollout_from_model = RolloutFromModel(model, ref_model, reward_model)
+
+    best_val_reward = float("-inf")
+    it = 0  # it is equivalent to batch_size number of episodes
+    with tqdm(total=int(max_epochs * num_rollouts_per_epoch / batch_size)) as pbar:
+        for _epoch in range(1, max_epochs + 1):
```
Comment (vmoens) [inline examples/rlhf/train_rlhf.py:258]: remove underscore

File: examples/rlhf/train_rlhf.py [hunk 10]
```diff
@@ -0,0 +1,339 @@
...
+            rb.empty()
+            rollout_rewards = []
+            rollout_kl = []
+            kl_controller = AdaptiveKLController(0.1, 6, 10000)
```
Comment (vmoens) [inline examples/rlhf/train_rlhf.py:262]: shouldn't this be created out of the loop?
We create it, update it, then override it with a new one

File: examples/rlhf/train_rlhf.py [hunk 11]
```diff
@@ -0,0 +1,339 @@
...
+                )
+                with torch.no_grad(), ctx:
+                    # moving this to within epoch
+                    adv_fn(td)
+                # it's possible we didn't fill the replay buffer in the last iteration if
+                # generation stopped early, so we empty first before repopulating
+                rb.extend(flatten_td(td))
+                done = td.get(("next", "done"))
+                next_reward = td.get(("next", "reward_raw"))[done]
+                next_kl = td.get(("next", "reward_kl"))[done]
```
Comment (vmoens) [inline examples/rlhf/train_rlhf.py:276]: this can be done more efficiently via
```suggestion
                td_done = td[done.view(td.shape)]
                next_reward = td_done.get(("next", "reward_raw"))
                next_kl = td_done.get(("next", "reward_kl"))
```

File: examples/rlhf/train_rlhf.py [hunk 12]
```diff
@@ -0,0 +1,339 @@
...
+                )
+                pbar.set_description(f"TRAIN: {it=}: {rollout_reward=:.4f}")
+
+            for batch in rb:
+                rb_ppo.empty()
+                rb_ppo.extend(batch)
+                for _ in range(ppo_num_epochs):  # PPO epochs
+                    optimizer.zero_grad()
+                    # why don't we optimize at each step? Is accumulating grads better?
+                    # usually more small steps is better than a giant one
```
Comment (vmoens) [inline examples/rlhf/train_rlhf.py:307]: remove comment
by the way: the reason why ppl decompose a single minibatch in multiple mini minibatches is simply for memory constrains and because PPO in rlhf requires big batch size
If a singel minibatch fits in memory we can perfectly do that

File: examples/rlhf/train_rlhf.py (hunk 4 above)
Comment (apbard) [reply examples/rlhf/train_rlhf.py:61]: it's just a scalar, no?

File: examples/rlhf/train_rlhf.py (hunk 6 above)
Comment (apbard) [reply examples/rlhf/train_rlhf.py:33]: done

File: examples/rlhf/train_rlhf.py (hunk 10 above)
Comment (apbard) [reply examples/rlhf/train_rlhf.py:262]: ops, sure

File: examples/rlhf/train_rlhf.py (hunk 8 above)
Comment (apbard) [reply examples/rlhf/train_rlhf.py:104]: to get a scalar instead of a scalar-tensor

File: examples/rlhf/train_rlhf.py (hunk 7 above)
Comment (apbard) [reply examples/rlhf/train_rlhf.py:67]: yes, we currently use that just to have a better estimate for logging purposes

File: torchrl/data/rlhf/utils.py [hunk 13]
```diff
@@ -87,7 +155,13 @@ class RolloutFromModel:
     EOS_TOKEN_ID = 50256
 
     def __init__(
-        self, model, ref_model, reward_model, max_new_tokens=50, score_clip=10.0
+        self,
+        model,
+        ref_model,
+        reward_model,
+        kl_controller,
```
Comment (vmoens) [inline torchrl/data/rlhf/utils.py:162]: this is missing from the doc.

I'm not so sure about this kl_controller that is passed to the module. I feel it should be handled separately. It's like passing the lr_scheduler to the optimizer, the reason we don't do that is that it mixes responsibilities between modules. It gives the impression that one module has multiple responsibilities but it is less clear than doing things explicitly in the main code.

Comment (apbard) [reply torchrl/data/rlhf/utils.py:162]: are you suggesting we go back to passing just the kl coefficient?

Comment (vmoens) [reply torchrl/data/rlhf/utils.py:162]: I think the KL controller should be another class, but then we have 2 options:
the KL controller changes the KL coefficienbt of the other class (like the LR scheduler changes the LR of the optimizer or the target param updaters in torchrl update the target params of the loss) or we explicitely pass the kl coef.
I think the first option is more "pytorch"-style

Comment (apbard) [reply torchrl/data/rlhf/utils.py:162]: fixed

File: examples/rlhf/train_rlhf.py [hunk 14]
```diff
@@ -0,0 +1,292 @@
...
+    )
+    rb_ppo = TensorDictReplayBuffer(
+        storage=LazyTensorStorage(episode_length * batch_size),
+        batch_size=ppo_batch_size,
+        sampler=SamplerWithoutReplacement(),
+        prefetch=10,
+    )
+
+    rollout_from_model = RolloutFromModel(model, ref_model, reward_model)
+    kl_controller = AdaptiveKLController(rollout_from_model, 0.1, 6, 10000)
```
Comment (vmoens) [inline examples/rlhf/train_rlhf.py:206]: I don't think AdaptiveKLController takes the rollout_from_model as input does it?

File: torchrl/data/rlhf/utils.py [hunk 15]
```diff
@@ -16,6 +18,72 @@
...
+        """Update ``self.coef`` adaptively.
+
+        Arguments:
+            kl_value: The current KL value between the newest policy and the initial
+                policy.
+            n_steps: The number of training steps taken since last update.
+        """
+        proportional_error = np.clip(kl_value / self.target - 1, -0.2, 0.2)  # ϵₜ
+        mult = 1 + proportional_error * n_steps / self.horizon
+        self.coef *= mult  # βₜ₊₁
```
Comment (vmoens) [inline torchrl/data/rlhf/utils.py:84]: Are these guys really part of data?
They seem more related to the model to me.
They act on a class that belongs to data (maybe should be moved to collector tbh) but the KL coef is something that has to do with the stochastic policy (the language model, in our case), not the data.

File: torchrl/data/rlhf/utils.py (hunk 13 above)
Comment (vmoens) [reply torchrl/data/rlhf/utils.py:162]: Should we not remove it then?

File: torchrl/data/rlhf/utils.py [hunk 16]
```diff
@@ -99,18 +173,19 @@ def __init__(
         self.reward_model = reward_model
         self.max_new_tokens = max_new_tokens
         self.score_clip = score_clip
+        self.kl_controller = kl_controller
 
-    def kl_step(self):
+    def kl_update(self, kl_value, n_steps):
         """Makes a step in the KL coefficient schedule."""
-        raise NotImplementedError
+        self.kl_controller.update(kl_value, n_steps)
```
Comment (vmoens) [inline torchrl/data/rlhf/utils.py:180]: ditto, maybe this function should go away?

File: torchrl/data/rlhf/utils.py [hunk 17]
```diff
@@ -167,7 +242,7 @@ def create_rollout_td(self, batch, generated, log_probs, log_ratio, kl_coef=0.1)
         )
         reward_raw = clipped_scores.unsqueeze(-1).unsqueeze(-1)
         reward_raw = reward_raw * done
-        reward_kl = -kl_coef * log_ratio.unsqueeze(-1)
+        reward_kl = -self.kl_controller.coef * log_ratio.unsqueeze(-1)
```
Comment (vmoens) [inline torchrl/data/rlhf/utils.py:245]: same here

File: torchrl/data/rlhf/utils.py (hunk 15 above)
Comment (vmoens) [reply torchrl/data/rlhf/utils.py:84]: New classes should be added to the doc (provided we're sure of where they belong)
